subscriptions-api.md

```diff
@@ -0,0 +1,613 @@
+hierarchy, and all current subscriptions whose topic filters match the path
+receive a message.
+
+The operations are normatively defined in the [MQTT
+3.1.1](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/mqtt-v3.1.1.html) and [MQTT
+5.0](http://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html) specifications.
+
+#### 3.1.2. AMQP 1.0
+
+The AMQP 1.0 protocol has a built-in subscription mechanism.
+
+When an AMQP container establishes a receive-role link to another container, it
+can specify a "distribution-mode" on the link source in the other container. If
+the distribution mode is "copy", each such established link receives a copy of
+the message stream available at the distribution node. If the distribution mode
+is "move", messages are exclusively owned by the link and are removed from the
+source when the transfer is settled successfully.
+
+Supporting the "copy" mode, is optional for the underlying implementation of the
+broker
+
+The link source can also be configured with filters. The AMQP core specification
+does not define concrete filter types, but the AMQP Filter Expressions 1.0
+extension specification and the Apache Qpid project's filter definitions do.
+
+#### 3.1.3. NATS
+
+The NATS protocol has a native subscribe operation "SUB" (and matching "UNSUB"),
+which allows a client to solicit a stream of messages that match a given subject
+or a wildcard.
+
+Events are published using the "PUB" operation with a subject name that is then
+matched against the existing subscriptions.
+
+#### 3.1.4. Apache Kafka
+
+Apache Kafka allows consumers to fetch events stored in topic partitions once or
+multiple times.
+
+While Apache Kafka does not have a server-side subscribe operation and also does
+not provide filtering capabilities, it allows the fetching client to fully
+control from which offset in the partition's log store events are being read
+from. This allows multiple parties to retrieve the same set of events from the
+stream.
+
+The management of the subscription and selecting which events are being fetched
+and dispatched into the application lies completely with the consumer in the
+Apache Kafka case.
+
+#### 3.1.5. HTTP
+
+HTTP does not have a built-in subscription mechanism that allows for
+establishing a flow of events similar to the aforementioned protocols. HTTP
+allows for modeling pull-style retrieval of events from a store and, with HTTP/2
+"server push", even for continuous delivery an event stream triggered by a
+initial request, but for event delivery scenarios, these techniques are
+applications of HTTP rather than inherent features of HTTP per-se.
+
+### 3.2. Subscription Manager API
+
+Subscriptions where the delivery is initiated by the subscription manager
+("push"-style) require a mechanism to configure such subscriptions on the
+subscription manager.
+
+The subscription manager API defines a subscription object and an API for
+creating, updating and deleting subscriptions on a subscription manager.
```

Comment:
IIUC, the motivational sentence here states that the mechanism (i.e. the protocol) was ultimately necessary in case of push-style delivery because the subscription embodies the consumer's interest in events and that lies initially with the consumer, so the SM needs to be notified of that interest. It's true, that especially "push"-style delivery requires an out-of-band mechanism to communicate a consumer's interest in events to the subscription manager.

```suggestion
The subscription manager API defines a subscription object and a protocol for 
creating, updating and deleting subscriptions on a subscription manager.

The subscription object describes the consumer's interest in events and 
defines the delivery method. The protocol allows to configure subscriptions
on the subscription manager. Especially in case of subscription manager 
initiated ("push"-style) delivery, the protocol is required to express the 
consumer's interest to the subscription manager before the subscription 
manager can initiate delivery.
```

But, wouldn't "pull"-style also require a mechanism to configure subscriptions at the subscription manager? IIUC a CloudEvent subscription manager deals in CloudEvents, i.e. routes them based on their event context. AFAICS no (mentioned) "pull"-style message delivery would **natively** support such CloudEvent context, e.g. where the consumer can initiate delivery of a specific CloudEvent type. The pulling consumer also requires a mechanism to have the subscription manager provide an endpoint/topic/queue that caters for the subscription.

The subscription manager is two things:
- the API endpoint/service that manages CRUD of declared CloudEvent Subscriptions
- the broker/router/producer/source/entity that delivers events push-style or serves interfaces for pull-style delivery

Native subscription management mechanisms may reuse the delivery transport channel to also manage e.g. topic subscriptions (in-band), but none of them talk _CloudEvent subscriptions_. So, some sort of out-of-band _subscription API_ is a necessity to communicate _CloudEvent subscription_ objects.

In general, isn't it that any CloudEvent subscription (push or pull), if expressed using this specification, always requires out-of-band communication to be setup? A consumer may use a native messaging "subscription" mechanism to _initiate delivery_, but still requires other means, e.g. to setup a topic and route messages to that topic according to the CloudEvent subscription it wants to realize.

Reply:
@manuelstein I've been thinking about these questions, too. I tried to describe some of my thoughts in the _Eventing domains_ sections in the original Google doc. 
The examples you mentioned seem to go beyond subscriptions as such. Isn't it more about setting up resources for routing or consumption? Is the configuration of a topic needed for subscribing or for producing? If you made this a part of subscribing, wouldn't you still require the consumer to do an additional subscription step based on the messaging protocol? 
A similar case is the creation of a queue that gets subscribed to the according topics. I think all this can be a layer on top of the raw subscription API we are defining here. 
If you like more discussions on this topic, just ping me on Slack.